### DISCUSSION TITLE
Setting AWS input files

### DISCUSSION THREAD

**Discussion by riazarbi** [ropensci/targets#315] [Help] [answered]
Hello, 

TL;DR - Is it currently possible to set a `tar_target` to read input data from an unsupported file sitting in AWS (parquet in my instance)?

I'm trying to set input file target that sits on aws s3. By way of example, if you take the canonical minimal example, 

```
# _targets.R
library(targets)
path_to_data <- function() {
  "data/raw_data.csv"
}
list(
  tar_target(
    raw_data_file,
    path_to_data(),
    format = "file"
  ),
  tar_target(
    raw_data,
    read_csv(raw_data_file, col_types = cols())
  )
)
```

I would like to modify the above code so that I am referencing an unsuported file format in AWS. Like so - 

```
# _targets.R
library(targets)
path_to_data <- function() {
  "raw_data.parquet"
}
list(
  tar_target(
    raw_data_file,
    path_to_data(),
    format = "aws_file",
   resources = list(bucket ="raw_data_bucket")
  ),
  tar_target(
    raw_data,
    read_parquet(raw_data_file)
  )
)
```

Now, I know that my `aws.s3` creds are working because I can use `aws_fst` as a `_targets` storage backend. But when I run the above code (or similar) I get -

```
● run target raw_data
x error target raw_data
Error : IOError: Failed to open local file 'raw_data.parquet'. Detail: [errno 2] No such file or directory
Error: callr subprocess failed: IOError: Failed to open local file 'data.parquet'. Detail: [errno 2] No such file or directory`
```

I can get the data by just using `aws.s3::save_object` in a `tar_target` but I lose out on all that "watch for upstream changes" goodness. 

So. Am I trying to do something that is not actually possible?

**Comment by riazarbi** [comment]
So I can see that if I create a file in my local environment called `data.parquet`, `tar_make()` on the above `_targets.R` file will take that file, upload it to `raw_data_bucket/_targets/objects/raw_data_file` and then monitor it for downstream activity.

But my use case is different - in my situation there is _already_ a parquet file in a bucket, which is refreshed periodically, that I want to read into my targets folder.

**Reply by wlandau** [reply]
You could track the [website endpoint](https://docs.aws.amazon.com/AmazonS3/latest/userguide/WebsiteEndpoints.html#WebsiteRestEndpointDiff) with `format = "url"` and then download it in a downstream target.

```r
# _targets.R
library(targets)
list(
  tar_target(
    aws_url,
    "http://bucket-name.s3-website.Region.amazonaws.com/folder-name/object-name.parquet",
    format = "url"
  ),
  tar_target(
    local_file, {
      download.file(aws_url, "local-data.parquet")
      "local-data.parquet"
    },
    format = "file"
  )
)
```

**Reply by riazarbi** [reply]
This occurred to me, but I was unsure how to deal with auth. These buckets are protected and you would need a key and secret to access them.

**Reply by riazarbi** [reply]
Ok, on the assumption that there is no convenience function to construct an S3 authenticated `curl` request in the targets package, I'll work on a function to construct one and post it here if I succeed.

**Reply by wlandau** [reply]
> This occurred to me, but I was unsure how to deal with auth. These buckets are protected and you would need a key and secret to access them.

Can these be set with environment variables? You could define them in a `.Renviron` file at the root of your project.

**Reply by riazarbi** [reply]
`.Renviron` is not a good solution for me as I'm going to be running `targets` workflows in disposable container workloads on Apache Airflow. Our usual route is to pass in a `secrets.json` file to the container and then have our scripts read them in at 
runtime.

I see that I might be an edge case here, but am reasonably sure I can come up with the solution myself, so I'll just post my progress here if that's ok? It might help someone else.

I've made some progress on writing a function that uses R `curl` to authenticate with a Minio (S3 compatible) server and download an object.

Here it is for posterity:

```
object_name = "REDACTED.parquet"
save_path = "REDACTED.parquet"
mimetype = "application/octet-stream"
bucket = "REDACTED"
minio_key = "REDACTED"
minio_secret = "REDACTED"
minio_url = "REDACTED"

minio_to_file <- function(object_name, 
                          save_path,
                          bucket,
                          mimetype,
                          minio_key,
                          minio_secret,
                          minio_url) {
  require(curl)
  require(digest)
  require(RCurl)
  object_name=object_name
  save_path=save_path
  contentType= mimetype
  bucket=bucket 
  dateValue=format(Sys.time(), "%Y%m%dT%H%M%SZ", tz = "UTC")
  stringToSign=paste0("GET\n\n",contentType,"\n",dateValue,"\n/",bucket,"/",object_name)
  s3Key=minio_key 
  s3Secret=minio_secret
  s3Host= minio_url
  signature <- base64(hmac(s3Secret, stringToSign,  algo = "sha1", raw = TRUE))
  
  url <- paste0("https://",s3Host,"/",bucket,"/",dest_path)
  headers <- list(Host = s3Host,
                  Date = dateValue,
                  `Content-Type` = contentType,
                  Authorization = paste0("AWS ",s3Key,":",signature))
  
  h <- new_handle()
  handle_setheaders(h, .list = headers)
  
  r <- curl_download(url, save_path, handle = h)

  
}

minio_to_file(object_name, 
                          save_path,
                          bucket,
                          mimetype,
                          minio_key,
                          minio_secret,
                          minio_url)

read_parquet(dst_proper_prefix_path)
unlink(dst_proper_prefix_path)


```

**Reply by riazarbi** [reply]
Ok, I've made some progress. I've whittled down the above to a function that accepts a url and returns an AWS signed header.

```
create_s3_handle <- function(object_url) {
  domain <-
    function(x)
      strsplit(gsub("http://|https://|www\\.", "", x), "/")[[c(1, 1)]]
  bucket <-
    function(x)
      strsplit(gsub("http://|https://|www\\.", "", x), "/")[[c(1, 2)]]
  object_name <- function(x) {
    y <- gsub("http://|https://|www\\.", "", x)
    z <-
      stringr::str_remove(y, paste0(domain(x), "/"))
      stringr::str_remove(z, paste0(bucket(x), "/"))
    
  }
  
  date_time <- format(Sys.time(), "%Y%m%dT%H%M%SZ", tz = "UTC")
  curl::handle_setheaders(
    curl::new_handle(),
    .list = list(
      Host = domain(object_url),
      Date = date_time,
      `Content-Type` = "application/octet-stream",
      Authorization = paste0(
        "AWS ",
        Sys.getenv("AWS_ACCESS_KEY_ID"),
        ":",
        RCurl::base64(digest::hmac(
          Sys.getenv("AWS_SECRET_ACCESS_KEY"),
          paste0(
            "GET\n\n",
            "application/octet-stream",
            "\n",
            date_time,
            "\n/",
            bucket(object_url),
            "/",
            object_name(object_url)
          ),
          algo = "sha1",
          raw = TRUE
        ))
      )
    )
  )

}

```

When I use the above function in a script, like so, I can curl in the file successfully from my local minio server - 

```
library(curl)
library(dplyr)
library(arrow)
save_path = "tmp.parquet"
object_url <-
  "https://HOST/BUCKET/tmp.parquet"

source("R/functions.R")

curl:::handle_getheaders(create_s3_handle(object_url))
#create_s3_handle(object_url)
tar_read(handle_to_inspect)

r <- curl_download(object_url, save_path, quiet = F, handle = create_s3_handle(object_url))

read_parquet(save_path) %>% glimpse()
unlink(save_path)
```

However, when I use the function in `tar_make`, I get an error:

```
> tar_make()
✓ skip target handle_to_inspect
● run target aws_url
Error : HTTP response status code 403
Could not access url:
  https://REDACTED/tmp.parquet
HTTP response headers:
  accept-ranges = bytes
  content-security-policy = block-all-mixed-content
  date = Thu, 18 Feb 2021 10:45:38 GMT
  server = MinIO
  vary = Origin
  x-amz-request-id = 1664D1B43A1758D2
  x-xss-protection = 1; mode=block
  content-length = 0
Error: callr subprocess failed: HTTP response status code 403
Could not access url:
  https://REDACTED/tmp.parquet
HTTP response headers:
  accept-ranges = bytes
  content-security-policy = block-all-mixed-content
  date = Thu, 18 Feb 2021 10:45:38 GMT
  server = MinIO
  vary = Origin
  x-amz-request-id = 1664D1B43A1758D2
  x-xss-protection = 1; mode=block
  content-length = 0
Type .Last.error.trace to see where the error occured
```

Here's the `_targets` file snippet:

```
# set input paths
path_to_aws_object <- "https://REDACTED/tmp.parquet"

list(
    tar_target(
      aws_url,
      path_to_aws_object,
      format = "url",
      resources = list(handle = create_s3_handle(path_to_aws_object))
    ),
   tar_target(
    handle_to_inspect, 
    curl:::handle_getheaders(create_s3_handle(path_to_aws_object)), 
    format = "rds"
  )
)
```

Interestingly enough, when I check the signatures of the header generated in the script vs the header generated in `tar_make`, they are the same. (In order to make sure they are hashing the same string I temporarily formatted out the milliseconds in the date signature):

```
> curl:::handle_getheaders(create_s3_handle(object_url))
[1] "Host: REDACTED"                                          
[2] "Date: 20210218T1050"                                                 
[3] "Content-Type: application/octet-stream"                              
[4] "Authorization: AWS MDDDOD2JZA43LJMLKTPM:h1XrlLeK6CnTLeEcin3uubyK1QY="
[5] "Expect: "                                                            

> #create_s3_handle(object_url)
> tar_read(handle_to_inspect)
[1] "Host: REDACTED"                                          
[2] "Date: 20210218T1050"                                                 
[3] "Content-Type: application/octet-stream"                              
[4] "Authorization: AWS MDDDOD2JZA43LJMLKTPM:h1XrlLeK6CnTLeEcin3uubyK1QY="
[5] "Expect:
```

My conclusions, therefore, are:

1. `tar_make` is generating the right handle
2. The handle is for some reason not being passed to the `url` call.

@wlandau does the above error spark any bright ideas? I'm pretty much tapped out.

**Reply by wlandau** [reply]
Could be the pipeline implicitly relies on something you wrote in your global `.Rprofile` or `.Renviron`? `tar_make()` runs through `callr`, which by default only pays attention to the local dotfiles in the working directory. If that explains it, `tar_make(callr_function = NULL)` avoids this (but is not recommended for serious runs). `callr_arguments` gives you some flexibility to use other dotfiles, or you could create copies local to the project.

**Reply by riazarbi** [reply]
Ok, @wlandau I went down a rabbit hole, poked around your source, and solved my problem. 

If I comment out line 50 of `utils_url.R`, then my `tar_make()` runs successfully.

```
url_handle <- function(handle = NULL) {
  handle <- handle %||% curl::new_handle()
  assert_inherits(handle, "curl_handle")
  #curl::handle_setopt(handle, nobody = TRUE)
  handle
}
```

It seems to me (and I could totally be wrong) that `curl::handle_setopt` will just replace those options named, while `curl::handle_setheaders` replaces the entire header set. 

So the headers I am passing in are not replacing the header set completely. 

I looked at trying to change this line of code `curl::handle_setopt(handle, nobody = TRUE)`, but then I noticed that you are actually using the `handle=handle` option pretty much everywhere in the script. 

So, my question to you - is that line actually needed? Maybe it can be deleted?

**Reply by riazarbi** [reply]
The way I see it, removing that line ensures that the handle sent in via the `resources` parameter will be all and only what gets passed to the `curl` call, which is desirable. 

One must just be sure that all `curl` calls in the package are accompanied by a `handle` parameter and preceded by a `url_handle` call.

**Reply by wlandau** [reply]
We only want the headers, and `nobody  = TRUE` ensures we do not accidentally download the whole body of the file ("no body"). I did not realize this would cause problems.

**Reply by wlandau** [reply]
I agree with your suggestion because it prevents surprises like the one you described. As of ae0f8f6d2e15313421d75ea9bdc753019685cc23, the default handle has `nobody = TRUE` and custom handles are not modified. The docs now urge people to set `nobody = TRUE` in their own custom handles.

**Reply by riazarbi** [reply]
Bellissimo!

**Comment by riazarbi** [comment, marked as answer]
Here, for posterity, is my setup. This allows me to use the "url" format on minio-hosted S3 objects. The url path construction may need to be modified for it to work with S3 subdomain-style bucket paths, YMMV. 

Will only work after https://github.com/ropensci/targets/commit/ae0f8f6d2e15313421d75ea9bdc753019685cc23.

It assumes that environment variables `AWS_ACCESS_KEY_ID`  and `AWS_SECRET_ACCESS_KEY` are set.

In `R/functions.R` 

```
create_s3_handle <- function(object_url) {
  domain <-
    function(x)
      strsplit(gsub("http://|https://|www\\.", "", x), "/")[[c(1, 1)]]
  bucket <-
    function(x)
      strsplit(gsub("http://|https://|www\\.", "", x), "/")[[c(1, 2)]]
  object_name <- function(x) {
    y <- gsub("http://|https://|www\\.", "", x)
    z <-
      stringr::str_remove(y, paste0(domain(x), "/"))
      stringr::str_remove(z, paste0(bucket(x), "/"))
    
  }
  
  date_time <- format(Sys.time(), "%Y%m%dT%H%M%SZ", tz = "UTC")
  curl::handle_setheaders(
    curl::new_handle(),
    .list = list(
      Host = domain(object_url),
      Date = date_time,
      `Content-Type` = "application/octet-stream",
      Authorization = paste0(
        "AWS ",
        Sys.getenv("AWS_ACCESS_KEY_ID"),
        ":",
        RCurl::base64(digest::hmac(
          Sys.getenv("AWS_SECRET_ACCESS_KEY"),
          paste0(
            "GET\n\n",
            "application/octet-stream",
            "\n",
            date_time,
            "\n/",
            bucket(object_url),
            "/",
            object_name(object_url)
          ),
          algo = "sha1",
          raw = TRUE
        ))
      )
    )
  )
}
```

My `_targets.R` file:

```
library(targets)

source("R/functions.R")

tar_option_set(
  packages = c(
    "arrow",
    "dplyr",
    "stringr",
    "curl",
    "digest",
    "RCurl"),
  format = "fst_tbl",
  memory = "persistent",
  error = "stop"
)


# set input paths
path_to_aws_object <- "https://MINIO_HOST/MINIO_BUCKET/tmp.parquet"

list(
  # example of a live make-like input conneciton
     tar_target(
       aws_url,
       path_to_aws_object,
       format = "url",
       resources = list(handle = create_s3_handle(path_to_aws_object))
     ),
   tar_target(
     parquet_df, {
       tf <- tempfile()
       curl_download(aws_url, 
                     tf, 
                     handle = create_s3_handle(aws_url)
                     )
       read_parquet(tf)
       unlink(tf)
       
     }
   )
)
```
